---
title: "scrap: Single-Cell Reproducibility Across Patients"
author: "Michael Cole"
date: "`r Sys.Date()`"
bibliography: bibFile.bib
output:
  BiocStyle::html_document:
    toc: true
vignette: >
  %\VignetteEncoding{UTF-8}
---

<!--
%\VignetteEngine{knitr::rmarkdown}
%\VignetteIndexEntry{scone Vignette}
-->

```{r options, include=FALSE, cache=FALSE, results='hide', message=FALSE}

knitr::opts_chunk$set(fig.align="center", cache=FALSE,error=FALSE,
                      fig.width=6,fig.height=6,autodep=TRUE,
                      out.width="600px", out.height="600px",
                      results="markup", echo=TRUE, eval=TRUE)

options(getClass.msg=FALSE)

set.seed(1776) ## for reproducibility

library(scrap)
library(SummarizedExperiment)
library(Rtsne)
library(RColorBrewer)
library(NMF)

```

# Introduction

## Loading Data

We are currently working from a personal path:

```{r datain}

# Load scRNA-Seq (Normalized and log-tranformed)
load(file = "~/Dropbox/2016 Gayo Single Cell DC HIV/Analysis/normleDat.Rda")

# Rename patients
colData(normleDat)$patient = c(1, 3, 2)[colData(normleDat)$patient]
colData(normleDat)$patient = factor(paste0("P",colData(normleDat)$patient))

## ------ Select 48 Hour Cells
normleDat_late = normleDat[, colData(normleDat)$timepoint == "48H"]
normleDat_late = normleDat_late[apply(assay(normleDat_late), 1, var) > 0, ]

```

# Clustering

## PAM Clustering via Dimension Reduction

```{r cluster}

pamkd_obj = pamkd(
  assay(normleDat_late),
  maxk = 10,
  maxd = 50,
  to_log = FALSE,
  pca_center = TRUE,
  pca_scale = TRUE
)

```

## Visualizing Results with tSNE

```{r tsne}

set.seed(1237)
tsne_obj = Rtsne(pamkd_obj$Y,
                        pca = FALSE,
                        max_iter = 5000,
                        verbose = FALSE)

```

```{r plottsne}

patient_pchs = c(21, 22, 24)
cluster_cols = c("turquoise1",
                 "darkolivegreen1",
                 "black",
                 "azure1",
                 "tomato1")

plot(
  tsne_obj$Y,
  xlim = c(-25, 25),
  ylim = c(-25, 25),
  pch = patient_pchs[as.numeric(colData(normleDat_late)$patient)],
  bg = cluster_cols[pamkd_obj$pam$clustering],
  main = "DC Expression State Manifold at 48H",
  xlab = "tSNE Dim 1",
  ylab = "tSNE Dim 2"
)

legend("topright",
       pch = patient_pchs,
       legend = paste0("P", 1:3))

o = c(1, 3, 5, 4, 2) # Alternative cluster labels
clustering = paste0("c", o[pamkd_obj$pam$clustering])

legend(
  "bottomright",
  fill = cluster_cols[order(o)],
  legend = paste0("c", 1:5)
)
  
```

# Modules

## Adjacency Matrix

```{r adj}

A = get.repro.thresh.adjacency(x = assay(normleDat_late),
                              r = colData(normleDat_late)$patient,
                              mad_constant = (1 / qnorm(3 / 4)) ^ 2,
                              var_thresh = 10 ^ -5 )

```

## Hub Identification

```{r hub}

is_hub = p.adjust(pzipdegree(igraph::graph.adjacency(A,
                                                     mode = "undirected")),
                  method = "bonferroni") < 0.01

# Compute hub correlations per Replicate
rep_cor_list = sapply(levels(colData(normleDat_late)$patient),
                      simplify = FALSE,
                      function(p) {
                        cor(t(assay(normleDat_late)[is_hub,
                              colData(normleDat_late)$patient == p]))
                        })

# Median hub correlations
mediancor = apply(simplify2array(rep_cor_list), 1:2, median)

```

## Module Clustering

```{r hubclust}

hc = hclust(dist(1 - cor(mediancor,
                         method = "pearson")),
            method = "complete")
hub_clustering = factor(paste0("m",cutree(hc, k = 3)))
module_cols <- brewer.pal(3, "Set1")


```

```{r mediancor}

aheatmap(
  main = "Median Hub-Gene Correlations",
  mediancor,
  breaks = 0,
  annCol = list(Modules = hub_clustering),
  annColors = list(Modules = module_cols),
  distfun = "correlation",
  hclustfun = "complete"
)


```

```{r patientcors}

hm_list = sapply(levels(colData(normleDat_late)$patient),
       function(p) {
         aheatmap(
           main = paste("Hub-Gene Correlations in",p),
           rep_cor_list[[p]],
           breaks = 0,
           annCol = list(Modules = hub_clustering),
           annColors = list(Modules = module_cols),
           distfun = "correlation",
           hclustfun = "complete"
         )
       }
)

```
# Session Info

```{r session}
sessionInfo()
```